File: dictionary-core/src/main/java/dev/bee/kanjianki/core/DictionaryTextUtil.java
package dev.bee.kanjianki.core;

import java.util.Locale;
import java.util.regex.Pattern;

final class DictionaryTextUtil {
    private static final Pattern MULTI_WHITESPACE = Pattern.compile("\\s+");

    private DictionaryTextUtil() {
    }

    static String stripHtml(String value) {
        if (value == null || value.isEmpty()) {
            return "";
        }
        return MULTI_WHITESPACE.matcher(htmlEntities(stripHtmlTags(value))).replaceAll(" ").trim();
    }

    static boolean isKanji(int cp) {
        return (cp >= 0x3400 && cp <= 0x4DBF)
                || (cp >= 0x4E00 && cp <= 0x9FFF)
                || (cp >= 0xF900 && cp <= 0xFAFF)
                || (cp >= 0x20000 && cp <= 0x2EBEF);
    }

    private static String stripHtmlTags(String value) {
        StringBuilder out = new StringBuilder(value.length());
        int index = 0;
        boolean done = false;
        while (!done && index < value.length()) {
            TagBounds tag = nextTag(value, index);
            if (tag.missingStart()) {
                out.append(value, index, value.length());
                done = true;
            } else {
                out.append(value, index, tag.start);
                if (tag.missingEnd()) {
                    out.append(value, tag.start, value.length());
                    done = true;
                } else {
                    index = appendTagReplacement(value, out, tag);
                }
            }
        }
        return out.toString();
    }

    private static TagBounds nextTag(String value, int index) {
        int tagStart = value.indexOf('<', index);
        int tagEnd = tagStart < 0 ? -1 : value.indexOf('>', tagStart + 1);
        return new TagBounds(tagStart, tagEnd);
    }

    private static int appendTagReplacement(String value, StringBuilder out, TagBounds tag) {
        if (tag.empty()) {
            out.append(value, tag.start, tag.afterEnd());
            return tag.afterEnd();
        }
        String tagName = openingTagName(value, tag.start + 1, tag.end);
        int skippedContentEnd = skippableContentEnd(value, tag.afterEnd(), tagName);
        if (skippedContentEnd >= 0) {
            appendSkipReplacement(out, tagName);
            return skippedContentEnd;
        }
        out.append(' ');
        return tag.afterEnd();
    }

    private static int skippableContentEnd(String value, int fromIndex, String tagName) {
        if ("rt".equals(tagName) || "style".equals(tagName) || "script".equals(tagName)) {
            return closingTagEnd(value, fromIndex, tagName);
        }
        return -1;
    }

    private static void appendSkipReplacement(StringBuilder out, String tagName) {
        if (!"rt".equals(tagName)) {
            out.append(' ');
        }
    }

    private static String openingTagName(String value, int index, int tagEnd) {
        char first = value.charAt(index);
        if (!Character.isLetter(first)) {
            return "";
        }
        int nameEnd = index + 1;
        while (nameEnd < tagEnd && isTagNameChar(value.charAt(nameEnd))) {
            nameEnd++;
        }
        return value.substring(index, nameEnd).toLowerCase(Locale.ROOT);
    }

    private static boolean isTagNameChar(char value) {
        return Character.isLetterOrDigit(value);
    }

    private static int closingTagEnd(String value, int fromIndex, String tagName) {
        String closingTag = "</" + tagName + ">";
        int maxStart = value.length() - closingTag.length();
        for (int index = fromIndex; index <= maxStart; index++) {
            if (value.regionMatches(true, index, closingTag, 0, closingTag.length())) {
                return index + closingTag.length();
            }
        }
        return -1;
    }

    private static String htmlEntities(String value) {
        return value
                .replace("&nbsp;", " ")
                .replace("&amp;", "&")
                .replace("&quot;", "\"")
                .replace("&#39;", "'")
                .replace("&lt;", "<")
                .replace("&gt;", ">");
    }

    private record TagBounds(int start, int end) {
        private boolean missingStart() {
            return start < 0;
        }

        private boolean missingEnd() {
            return end < 0;
        }

        private boolean empty() {
            return end == start + 1;
        }

        private int afterEnd() {
            return end + 1;
        }
    }
}

File: dictionary-core/src/main/kotlin/dev/bee/kanjianki/core/DictionaryTextUtil.kt
package dev.bee.kanjianki.core

import java.util.Locale
import java.util.regex.Pattern

internal object DictionaryTextUtil {
    private val MULTI_WHITESPACE: Pattern = Pattern.compile("\\s+")

    @JvmStatic
    fun stripHtml(value: String?): String {
        if (value.isNullOrEmpty()) {
            return ""
        }
        return MULTI_WHITESPACE.matcher(htmlEntities(stripHtmlTags(value))).replaceAll(" ").trim()
    }

    @JvmStatic
    fun isKanji(cp: Int): Boolean {
        return (cp in 0x3400..0x4DBF) ||
            (cp in 0x4E00..0x9FFF) ||
            (cp in 0xF900..0xFAFF) ||
            (cp in 0x20000..0x2EBEF)
    }

    private fun stripHtmlTags(value: String): String {
        val out = StringBuilder(value.length)
        var index = 0
        var done = false
        while (!done && index < value.length) {
            val tag = nextTag(value, index)
            if (tag.missingStart()) {
                out.append(value, index, value.length)
                done = true
            } else {
                out.append(value, index, tag.start)
                if (tag.missingEnd()) {
                    out.append(value, tag.start, value.length)
                    done = true
                } else {
                    index = appendTagReplacement(value, out, tag)
                }
            }
        }
        return out.toString()
    }

    private fun nextTag(value: String, index: Int): TagBounds {
        val tagStart = value.indexOf('<', index)
        val tagEnd = if (tagStart < 0) -1 else value.indexOf('>', tagStart + 1)
        return TagBounds(tagStart, tagEnd)
    }

    private fun appendTagReplacement(value: String, out: StringBuilder, tag: TagBounds): Int {
        if (tag.empty()) {
            out.append(value, tag.start, tag.afterEnd())
            return tag.afterEnd()
        }
        val tagName = openingTagName(value, tag.start + 1, tag.end)
        val skippedContentEnd = skippableContentEnd(value, tag.afterEnd(), tagName)
        if (skippedContentEnd >= 0) {
            appendSkipReplacement(out, tagName)
            return skippedContentEnd
        }
        out.append(' ')
        return tag.afterEnd()
    }

    private fun skippableContentEnd(value: String, fromIndex: Int, tagName: String): Int {
        if ("rt" == tagName || "style" == tagName || "script" == tagName) {
            return closingTagEnd(value, fromIndex, tagName)
        }
        return -1
    }

    private fun appendSkipReplacement(out: StringBuilder, tagName: String) {
        if ("rt" != tagName) {
            out.append(' ')
        }
    }

    private fun openingTagName(value: String, index: Int, tagEnd: Int): String {
        val first = value[index]
        if (!Character.isLetter(first)) {
            return ""
        }
        var nameEnd = index + 1
        while (nameEnd < tagEnd && isTagNameChar(value[nameEnd])) {
            nameEnd++
        }
        return value.substring(index, nameEnd).lowercase(Locale.ROOT)
    }

    private fun isTagNameChar(value: Char): Boolean = Character.isLetterOrDigit(value)

    private fun closingTagEnd(value: String, fromIndex: Int, tagName: String): Int {
        val closingTag = "</$tagName>"
        val maxStart = value.length - closingTag.length
        for (index in fromIndex..maxStart) {
            if (value.regionMatches(index, closingTag, 0, closingTag.length, true)) {
                return index + closingTag.length
            }
        }
        return -1
    }

    private fun htmlEntities(value: String): String {
        return value
            .replace("&nbsp;", " ")
            .replace("&amp;", "&")
            .replace("&quot;", "\"")
            .replace("&#39;", "'")
            .replace("&lt;", "<")
            .replace("&gt;", ">")
    }

    private class TagBounds(val start: Int, val end: Int) {
        fun missingStart(): Boolean = start < 0

        fun missingEnd(): Boolean = end < 0

        fun empty(): Boolean = end == start + 1

        fun afterEnd(): Int = end + 1
    }
}
